src/rules/ResourceXML.js

```diff
@@ -0,0 +1,202 @@
+        const suffix = '</root>';
+
+        // convert HTML single tags like <p> to <p/> and <br> to <br/> so
+        // that the xml parser doesn't freak out. But, we should only do
+        // this if the closing tags </p> and </br> do not already exist
+        // in the string
+        this.endTagRe.lastIndex = 0;
+        this.selfClosingRe.lastIndex = 0;
+        if (!this.endTagRe.test(source)) {
+            source = source.replace(this.selfClosingRe, "<$1/>");
+        }
+        const wrappedSource = `${prefix}${source}${suffix}`;
+
+        try {
+            srcObj = xml2js(wrappedSource, {
+                trim: false
+            });
+        } catch (e) {
+            // source is not well-formed, so don't even
+            // attempt to parse the target!
+            return undefined;
+        }
+
+        unnamedTagRe.lastIndex = 0;
+        if ((match = unnamedTagRe.exec(target)) !== null) {
+            const highlight =
+                target.replace(/</g, "&lt;").replace(/(&lt;\/?>)/g, "<e0>$1</e0>");
+            let opts = {
+                severity: "error",
+                rule: this,
+                description: `Empty XML elements <> and </> are not allowed in the target.`,
+                id: resource.getKey(),
+                highlight: `Target: ${highlight}`,
+                pathName: resource.getPath(),
+                source: resource.getSource(),
+                locale: resource.getTargetLocale()
+            };
+            problems.push(new Result(opts));
+        }
+
+        this.endTagRe.lastIndex = 0;
+        this.selfClosingRe.lastIndex = 0;
+        if (!this.endTagRe.test(target)) {
+            target = target.replace(this.selfClosingRe, "<$1/>");
+        }
+        const wrappedTarget = `${prefix}${target}${suffix}`;
+        try {
+            // then parse the target
+            tgtObj = xml2js(wrappedTarget, {
+                trim: false
+            });
+
+            problems = problems.concat(this.matchElements(srcObj, tgtObj, resource));
+        } catch (e) {
+            const lines = e.message.split(/\n/g);
+            // find the column number in the 3rd line of the exception message and subtract off
+            // the length of the prefix text we added in wrappedTarget
+            const column = parseInt(lines[2].substring(8)) - prefix.length;
+            // create the highlight, but make sure to escape any less than characters so that
+            // it does not conflict with the highlight
+            const highlight = column >= target.length ?
+                target.replace(/</g, "&lt;") + '<e0/>' :
+                target.substring(0, column).replace(/</g, "&lt;") + '<e0>' + target[column] + '</e0>' + target.substring(column+1).replace(/</g, "&lt;");
+            let opts = {
+                severity: "error",
+                rule: this,
+                description: `XML in translation is not well-formed. Error: ${lines[0]}`,
```

Comment:
It would be nice to make this less convoluted as well, such that the flow in code matches how you would do this by hand e.g. first massage unclosed tags in source and in target, then check source and target well-formedness, then check for unnamed tags and finally count named tags in source and target and compare the counts

Reply:
The order of the checks is as intended -- first you have to check the source for well-formedness. If it is not, then bail on all other checks. It is the job of a different rule to report on non-well-formedness of source strings and that rule should be used with source code linting.

Then, we have to check for the unnamed tags before we parse the target or else we will get errors that are mysterious: "unclosed tag" instead of "unnamed tag". Then we can parse the target string. If the parsing works, then, finally, we can match the source and target.

I'll write some better comments to explain this